### PR TITLE
Siamese layer

### PR DESCRIPTION
**Pull request by farizrahman4u**
`Siamese` layer is used for obtaining 2 or more identical branches whose weights are synchronized.

### CODE REVIEW

**Comment by farizrahman4u** [comment]
Siamese Layer

**Comment by farizrahman4u** [comment]
* Using a Siamese layer, you can effectively have **multiple identical layers with shared weights**. 
* The Siamese layer takes n inputs and gives n outputs
* The outputs from weight synchronized layers are merged using a specified merge_mode, similar to Merge class.
* Though thinking in terms of multiple layers with shared weights makes it easier to under stand, the actual implementation is different

siamese

**API Usage - (Objects named according to the above image)**
```
input_model_1 = Sequential()
input_model_1.add(Dense(input_dim=10, output_dim=10)) 

#input model 2
input_model_2 = Sequential()
input_model_2.add(Dense(input_dim=10, output_dim=10))

inputs = [input_model1, input_model2]

layer=Dense(input_dim=10, output_dim=5)

model = Sequential()
model.add(Siamese(layer ,inputs, 'sum'))

model.compile(loss='mse', optimizer='sgd')
```

**Note:**
* The weight shared layer could also be a model.

**Comment by farizrahman4u** [comment]
* Using a `Siamese` layer, you can effectively have **identical layers with shared weights**. 
* The `Siamese` layer takes 2 or more models as inputs. Each input is forwarded to their corresponding weight sync `Layer` (the orange ones).
* The outputs from the weight synchronized layers are merged using a specified `merge_mode`, similar to Merge class.
* Though thinking in terms of multiple layers with shared weights makes it easier to understand, the actual implementation is different.

Siamese

**API Usage - (Objects named according to the above image)**
```
input_model_1 = Sequential()
input_model_1.add(Dense(input_dim=10, output_dim=10)) 

input_model_2 = Sequential()
input_model_2.add(Dense(input_dim=10, output_dim=10))

inputs = [input_model1, input_model2]

layer = Dense(input_dim=10, output_dim=5)

model = Sequential()
model.add(Siamese(layer ,inputs, 'sum'))

model.compile(loss='mse', optimizer='sgd')
```

**Note:**
* The weight shared layers could also be  models.
* This is different from `Merge([leg,leg])`. Here you could chose at what level the layers of the legs should be in sync. Whereas in the case of `Merge[leg,leg]` , the legs are in sync down to the input level
* In other words, using this layer you could add synchronized layer(s) on on top of any two (or more) arbitrary models.

**Comment by farizrahman4u** [comment]
From #923 :
@sergeyf 
> The get_siamese_heads function feels a bit tacked on, but I am not sure how else to get it to do what it's doing without some changes to the base model class. I assume you are trying to avoid that?

Yes, obviously. There were some serious discussions going on at #362 , but they all involved drastic changes to the fundamental APIs, which I am not a fan of.

**Comment by sergeyf** [comment]
Repasting from another thread:

It looks very flexible and useful. The get_siamese_heads function feels a bit tacked on, but I am not sure how else to get it to do what it's doing without some changes to the base model class. I assume you are trying to avoid that?

Also, how does this fit in with the Graph api? Probably some of the more regular contributors to Keras would have a lot more to say, but as it is I would find immediate use for it, and so would many others I bet.

**Comment by farizrahman4u** [comment]
@sergeyf 
> The get_siamese_heads function feels a bit tacked on, but I am not sure how else to get it to do what it's doing without some changes to the base model class. I assume you are trying to avoid that?
Yes, obviously. There were some serious discussions going on at #362 , but they all involved drastic changes to the fundamental APIs, which I am not a fan of.
>Also, how does this fit in with the Graph api?
When `merge_mode='join`, the `Siamese` layer outputs an OrderedDict(), with keys `head1`, `head2` ..etc. This should be compatible with graph.

**Comment by farizrahman4u** [comment]
@sergeyf  Do you have any suggestions about the `get_siamese_heads` API?

**Comment by sergeyf** [comment]
Maybe something like:

```
input1 = Sequential()
input1.add(Dense(input_dim=10, output_dim=10))

input2 = Sequential()
input2.add(Dense(input_dim=10, output_dim=10))

inputs=[input1,input2]

shared_lstm = LSTM(input_dim=10, output_dim=5)

S = Siamese(shared_lstm ,inputs, 'sum',branch_names=['left','right'])

S.add_branch(branch_name='left',Dense(5))
S.add_branch(branch_name='right',LSTM(5))
```
But now it's starting to look like a `Graph` =)

**Comment by farizrahman4u** [comment]
@sergeyf  But wont this get complicated when we stack a `Siamese` over `Siamese` like in the second image?

**Comment by sergeyf** [comment]
Yeah...

Hmm what about explicitly constructing two branches and then indicating where they should be shared? That is a very different type of API, but it would look like this:

```
branch1 = Sequential()
branch1.add(Dense(10,10))
branch1.add(LSTM(10))
branch1.add(Dense(10))
branch1.add(Dense(1))

branch2 = Sequential()
branch1.add(Dense(10,10))
branch1.add(Dense(10))
branch1.add(Dense(10))
branch1.add(Dense(1))

model = Sequential(inputs=[branch1,branch2],merge_mode='sum',shared_tuples=[(branch1.layers[0],branch2.layers[0]),(branch1.layers[2],branch2.layers[2]))
```
Or simply by index:
`model = Sequential(inputs=[branch1,branch2],merge_mode='sum',shared_indices=[0,2])`
Which tells Sequential that the 1st and 3rd layer in each branch should be tied. Of course if they are different sizes, then it would throw an error, etc. But this would even let the branches be different lengths as long as the linked-up layers are actually identical.

**Comment by farizrahman4u** [comment]
Hmm. The simply by index API sounds good.. But again, under the hood, I am not explicitly typing up/sharing anything.. Because it is almost impossible due to theano issues and learning issues.

**Comment by sergeyf** [comment]
Yes, and it should be easy to also adapt to Graph then: just use tuples of
node names instead of tuples of indices.

**Comment by dbonadiman** [comment]
I find these utility layers really useful however from my point of view they are really messy in a sequential Network of sort.
I think that they are more natural in a Graph based network. In that sense they are already implementable without much effort, it is just to share the weight between standard layers in a Graph container. Then is possible to broadcast two different input to the two(or more) layers and the Graph does the trick.

I think that the Graph model was presented to avoid the implementation of these layers.

**Comment by farizrahman4u** [comment]
To all graph lovers, I have updated my first comment.

**Comment by sergeyf** [comment]
Thanks @farizrahman4u! With the `Graph` support, can you add multiple shared nodes in a row? So that there is a kind of merge_mode = `don't merge yet', forcing the next added layer also as a shared one?

**Comment by farizrahman4u** [comment]
@sergeyf  Interesting question. Simply set `merge_mode='join'`, so no actual merging will take place..the outputs will be simply bundled in an `OrderedDict`.

**Comment by farizrahman4u** [comment]
Hi @sergeyf , I have added support for stacking shared layers in `Graph`. Simply set `merge_mode=None` and `input=<name of previous shared layer>`

**Comment by sergeyf** [comment]
Sounds great!

**Comment by farizrahman4u** [comment]
If we use a separate argument `input` for stacking, we can not stack a shared layer on multiple stacks of shared layers like this:

```        
                       output
              ------ Merge------
           ---------------------------
           ---------------------------
           ---------------------------
           ---------           ---------
           ---------           ---------
           ---------           ---------
           in1,in2            in3,in4


```
So I am getting rid of the `input` argument.. the `inputs` argument can be used for stacking a shared layer on another shared layer (or multiple shared layers)..

**Comment by fchollet** [comment]
I do not understand the need for two APIs in the case of `Sequential`. Wouldn't `get_siamese_heads` be sufficient to cover all user needs?

**Comment by farizrahman4u** [comment]
@fchollet , Yes, it is sufficient..but in cases where the outputs from the shared layers are merged immediately,(the case where you dont want to stack any more shared layers on top) the first API can behave more like a `Merge` layer.. it takes in two models, and you add that to the third model..whereas if you use `get_siamese_heads` for the same, it takes in 2 models, outputs 2 models, and then you have to merge them to another model..

**Comment by farizrahman4u** [comment]
Again, you can not remove the merge functionality in `Siamese` layer, because it is used by the `add_shared_node()`

**Comment by farizrahman4u** [comment]
As the `get_siamese_heads()` function introduced too many models to the code, I have replaced it with  `add_shared_layer()`. It doesnt return new models, instead adds shared layer to the input models. And the name is more consistent with 'add_shared_node()`

**Comment by mmmikael** [comment]
What about this implementation that does not require any weight duplication:
https://github.com/fchollet/keras/issues/242#issuecomment-114996378

I have tested that on MNIST here https://gist.github.com/mmmikael/b202b344b1c89dbbce64.
It may need some more abstraction but it seems simpler to me.

**Comment by farizrahman4u** [comment]
@mikekestemont  Isnt that too specific to a particular problem? I am trying to do a generalized solution, where any layer could be added as a shared layer, without disturbing the way we train  the model..

**Comment by mmmikael** [comment]
> @mikekestemont Isnt that too specific to a particular problem? I am trying to do a generalized solution, where any layer could be added as a shared layer, without disturbing the way we train the model..

Yes, this is indeed specific but I was thinking that one could build a more generic implementation based on that idea. A ```merge``` could simply be implemented in a specific layer that reduces the batch size by a factor 2.

**Comment by farizrahman4u** [comment]
@mikekestemont  Can i see some pseudo code demonstrating how the API would look like? For e.g, the 2nd model in my first comment.. what would it look like ?

**Comment by mmmikael** [comment]
>  Can i see some pseudo code demonstrating how the API would look like? For e.g, the 2nd model in my first comment.. what would it look like ?

It could be something like:
```python
g = Graph()
g.add_input(name='input_pairs', input_shape=input_shape)
g.add_node(name='shared_dense_1', input='input_pair', Dense())
g.add_node(name='left_dense_1', input='shared_dense_1', Dense(), filter='even')
g.add_node(name='right_lstm_1', input='shared_dense_1', LSTM(), filter='odd')
g.add_node(name='shared_dense_2', inputs=['left_dense_1', 'right_lstm_1'], Dense(), merge_mode='batch_merge')
g.add_node(name='left_dense_2', input='shared_dense_2', Dense(), filter='even')
g.add_node(name='right_dense', input='shared_dense_2', Dense(), filter='odd')
g.add_node(name='merge', inputs=['left_dense_2', 'right_dense'], Merge(), merge_mode='batch_merge')
```
A ```filter``` mode would select the correct elements in the batch and the ```batch_merge``` merge mode would pick the correct elements for its two inputs (even elements from first inputs and odd from second).

**Comment by farizrahman4u** [comment]
@mmmikael I am not sure if this makes things easier or simpler. I will look into it. Thanks!

**Comment by farizrahman4u** [comment]
It seems the current implementation is pretty straight forward, you use `add_shared_layer()` in `Sequential` and `add_shared_node` in `Graph`.

**Comment by farizrahman4u** [comment]
@fchollet  It doesnt matter if there are 2 APIs for `Sequential`. Even if we dont' call the first method an "API", there would not be any change in the code. There would be a difference only in the documentation.

**Comment by farizrahman4u** [comment]
Travis CI error

**Comment by benjaminklein** [comment]
Are you planning to add an example for this? Thanks!

**Comment by farizrahman4u** [comment]
@benjaminklein  Consider the case where you are asked to make a model, which when given 2 sentences, gives a number representing how similar they are. You would:

```python

s1 = Sequential()
s1.add(Embedding(...))
s1.add(LSTM(...))
s1.add(Dense(...))

s2 = Sequential()
s2.add(Embedding(...))
s2.add(LSTM(...))
s2.add(Dense(...))

model = Sequential()
model.add(Merge([s1,s2], 'dot'))
model.add(Dense(1))
model.add(Activation('sigmoid'))
```
The above model has the following problems:
* **Order sensitive**: `similarity('hai', 'hello')` will be different from `similarity('hello','hai')`. Forcing you to iterate over your training set multiple times, each time shuffling your inputs..i.e, you would have to train ['hai', 'hello', 0.9] as well as ['hello', 'hai', 0.9]. Which means 2x training time. Now if your model have n inputs, you would have to iterate over your training set factorial(n) times to make your model order insensitive.
* **Model is heavy**: You have multiple layers for doing the same thing, that is learn phrase representation. This leads to redundancy, which means your model will have higher memory requirements and heavier file size(when you serialize/save weights).

**Solution**

```python
s1 = Sequential()
s2 = Sequential()

#add shared layers
add_shared_layer([s1,s2], Embedding(...))
add_shared_layer([s1,s2], LSTM(...))
add_shared_layer([s1,s2], Dense(...))

#everything else as usual
model = Sequential()
model.add(Merge([s1,s2], 'dot'))
model.add(Dense(1))
model.add(Activation('sigmoid'))
```
Hope this helps!

**Comment by farizrahman4u** [comment]
@fchollet What are your final thoughts about this? I have replaced the `get_siamese_heads` function with the more elegant `add_shared_layer`. Graph is also supported.

**Comment by fchollet** [comment]
I'm going mostly offline for 3 days, I'll review and answer this weekend.

My initial feeling is that we are likely to merge this, as it provides
useful functionality. UX tweaking might or might not be needed.

**Comment by farizrahman4u** [comment]
@fchollet Thanks..Once you are OK with the API, let me know so that I can start the documentation.

**Comment by dbonadiman** [comment]
@farizrahman4u @fchollet Guys, i was just wondering how can i use this layer in my code and i found another use case that maybe require some modification before merging. Can this kind of structure being used to implement the famous (RecurrentContainer/TimeDistributedModel) as soon as we have a fixed size input_lenght so that we can unroll the time dimension? 

In particular to you @farizrahman4u that you know better the limitation of this model.

**Comment by farizrahman4u** [comment]
@dbonadiman Please see my reply to @benjaminklein above. Its a real example of where this layer could be used.

**Comment by benjaminklein** [comment]
@farizrahman4u Thank you for the example!

File: keras/layers/containers.py [hunk 1]
```diff
@@ -277,6 +277,64 @@ def add_node(self, layer, name, input=None, inputs=[],
         if create_output:
             self.add_output(name, input=name)
 
+    def add_shared_node(self, layer, name, inputs=[], merge_mode=None, concat_axis=-1, dot_axes=-1, outputs=[], create_output=False):
```
Comment (fchollet) [inline keras/layers/containers.py]: Please add a docstring detailing the meaning of each argument. It is non-obvious.

File: keras/layers/core.py [hunk 2]
```diff
@@ -940,3 +940,229 @@ def get_config(self):
...
+            if len(shape) == 1:
+                shape.append(1)
+            return tuple(shape)
+        elif self.merge_mode == 'cos':
+            return tuple(input_shapes[0][0], 1)
+
+    def get_params(self):
+        return self.params, self.regularizers, self.constraints, self.updates
+
+    def get_output_at(self, head, train=False):
```
Comment (fchollet) [inline keras/layers/core.py]: Docstring please.

File: keras/layers/containers.py [hunk 3]
```diff
@@ -277,6 +277,64 @@ def add_node(self, layer, name, input=None, inputs=[],
...
+                raise Exception('Duplicate node identifier: ' + o)
+        if merge_mode:
+            if merge_mode not in {'sum', 'ave', 'mul', 'dot', 'cos', 'concat', 'join'}:
+                raise Eception("Invalid merge mode")
+        layers = []
+        for i in range(len(inputs)):
+            input = inputs[i]
+            if input in self.nodes:
+                n = self.nodes[input]
+                if hasattr(n, 'get_output_at'):#is it a siamese layer?
```
Comment (fchollet) [inline keras/layers/containers.py]: `if n.__class__.__name__ == ...`

File: keras/layers/core.py [hunk 4]
```diff
@@ -940,3 +940,229 @@ def get_config(self):
                   "input_dim": self.input_dim}
         base_config = super(MaxoutDense, self).get_config()
         return dict(list(base_config.items()) + list(config.items()))
+
+
+class Siamese(Layer):
+    def __init__(self, layer, inputs, merge_mode=None, concat_axis=1, dot_axes=-1):
```
Comment (fchollet) [inline keras/layers/core.py]: Docstring please.

File: keras/layers/core.py [hunk 5]
```diff
@@ -940,3 +940,229 @@ def get_config(self):
...
+
+        config = {"name": self.__class__.__name__,
+                  "layer": self.layer.get_config,
+                  "inputs": [m.get_config() for m in self.inputs],
+                  "merge_mode": self.merge_mode,
+                  "concat_axis": self.concat_axis,
+                  "dot_axes": self.dot_axes
+                  }
+        base_config = super(Siamese, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
```
Comment (fchollet) [inline keras/layers/core.py]: Line break here.

File: keras/layers/core.py [hunk 6]
```diff
@@ -940,3 +940,229 @@ def get_config(self):
...
+                  "head": self.head
+                  }
+
+        base_config = super(SiameseHead, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    def set_previous(self, layer):
+        self.previous = layer
+
+def add_shared_layer(layer,inputs):
```
Comment (fchollet) [inline keras/layers/core.py]: PEP8: space after comma

File: keras/layers/core.py [hunk 7]
```diff
@@ -940,3 +940,229 @@ def get_config(self):
...
+        config = {"name": self.__class__.__name__,
+                  "layer": self.layer.get_config,
+                  "inputs": [m.get_config() for m in self.inputs],
+                  "merge_mode": self.merge_mode,
+                  "concat_axis": self.concat_axis,
+                  "dot_axes": self.dot_axes
+                  }
+        base_config = super(Siamese, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+class SiameseHead(Layer):
```
Comment (fchollet) [inline keras/layers/core.py]: Docstring please.

File: keras/layers/core.py [hunk 8]
```diff
@@ -940,3 +940,229 @@ def get_config(self):
...
+    def __init__(self, head):
+        self.head = head
+        self.params = []
```
Comment (fchollet) [inline keras/layers/core.py]: Line break here.

**Comment by farizrahman4u** [comment]
On it. Need some time.

**Comment by farizrahman4u** [comment]
@fchollet  Before doing the doc strings and docs, it would be great if you could confirm the current `add_shared_node` API.

As of now, `add_shared_node` could be used in 3 ways:

* 1) Merge immediately after shared layer:
```python
graph = Graph()

graph.add_input(Dense(...), 'input1')
graph.add_input(Dense(...), 'input2')

graph.add_shared_node(Dense(...), name='shared_dense', inputs=['input1', 'input2'], merge_mode='concat', create_output=True)
```

* 2) Stack shared layers:

```python
graph = Graph()

graph.add_input(Dense(...), 'input1')
graph.add_input(Dense(...), 'input2')

graph.add_shared_node(Dense(...), name='shared1', inputs=[input1, input2])
graph.add_shared_node(Dense(...), name='shared2', inputs=[shared1])
graph.add_shared_node(Dense(...), name='shared3', inputs=[shared2])
graph.add_shared_node(Dense(...), name='shared4', inputs=[shared3])

graph.add_shared_node(Dense(...), name='merge', inputs=[shared4], merge_mode='sum', create_output=True)
```

* 3) Add normal layers on top of outputs from shared layer (Multi I/O mode)

```python
graph = Graph()

graph.add_input(Dense(...), 'input1')
graph.add_input(Dense(...), 'input2')

#Shared dense layer: takes 2 inputs, passes them through the dense layer and produces 2 outputs.
graph.add_shared_node(Dense(...), name='shared1', inputs=[input1, input2], outputs = ['output1','output2'])

#Add normal layers on top of shared layer outputs
graph.add_node(Dense(...), name='dense1', input='output1', create_output=True)
graph.add_node(Dense(...), name='dense2', input='output2', create_output=True)
```

**Comment by dbonadiman** [comment]
@farizrahman4u really well done i see lots of potential with this shared node class.
case 3 works with any number of layers or it needs max 2 input and max 2 output?

**Comment by EderSantana** [comment]
Sorry for being really late to comment here. Quick question @farizrahman4u . Are you concatenating the two (or more) branches in the batch dimension and doing just one `get_output`. Would that be bad? I know that it would break things if one leg is tensor3 and the other a matrix, still, there are very few layer (if any) that can take both cases as input.

**Comment by farizrahman4u** [comment]
@dbonadiman It can work work with as many inputs as you want:)

**Comment by farizrahman4u** [comment]
@EderSantana I think that would break everything. All the inputs to a shared layer should have identical shapes. There maybe layers which could eat both 2D and 3D data, but you can not make the same layer do both simultaneously. The `output_shape` of the shared layer is calculated from the `output_shape` of the first input. 
```
@property
def output_shape(self):
    self.layer.previous = inputs[0]#Assumption: All inputs have same shape. So return shape of first input
    return self.layer.get_output()
```
Hope this makes sense.

**Comment by bkj** [comment]
@farizrahman4u Is there an end-to-end example using this architecture on dummy data / one of the Keras datasets somewhere?

**Comment by farizrahman4u** [comment]
@bkj I will be adding a test as part of this PR soon. In the mean time, check out my reply to @benjaminklein above^ .

**Comment by farizrahman4u** [comment]
@fchollet  Before doing the doc strings and docs, it would be great if you could confirm the current `add_shared_node` API.

As of now, `add_shared_node` can be used in 3 ways:

* 1) Merge immediately after shared layer:
```python
graph = Graph()

graph.add_input(Dense(...), 'input1')
graph.add_input(Dense(...), 'input2')

graph.add_shared_node(Dense(...), name='shared_dense', inputs=['input1', 'input2'], merge_mode='concat', create_output=True)
```

* 2) Stack shared layers:

```python
graph = Graph()

graph.add_input(Dense(...), 'input1')
graph.add_input(Dense(...), 'input2')

graph.add_shared_node(Dense(...), name='shared1', inputs=[input1, input2])
graph.add_shared_node(Dense(...), name='shared2', inputs=[shared1])
graph.add_shared_node(Dense(...), name='shared3', inputs=[shared2])
graph.add_shared_node(Dense(...), name='shared4', inputs=[shared3])

graph.add_shared_node(Dense(...), name='merge', inputs=[shared4], merge_mode='sum', create_output=True)
```

* 3) Add normal layers on top of outputs from shared layer (Multi I/O mode)

```python
graph = Graph()

graph.add_input(Dense(...), 'input1')
graph.add_input(Dense(...), 'input2')

#Shared dense layer: takes 2 inputs, passes them through the dense layer and produces 2 outputs.
graph.add_shared_node(Dense(...), name='shared1', inputs=['input1', 'input2'], outputs = ['output1','output2'])

#Add normal layers on top of shared layer outputs
graph.add_node(Dense(...), name='dense1', input='output1', create_output=True)
graph.add_node(Dense(...), name='dense2', input='output2', create_output=True)
```

**Comment by fchollet** [comment]
Yes, I think the API makes perfect sense. To clarify, is this just for the `Graph` model?

**Comment by farizrahman4u** [comment]
@fchollet Yes. For `Sequential` its  `add_shared_layer`. Example:

```
input1=Sequential()
input2=Sequential()

input1.add(....)
input1.add(....)

input2.add(....)
input2.add(....)

add_shared_layer(Embedding(...), [input1,input2])
```

**Comment by fchollet** [comment]
Cool, this is very reasonable. The usage would need to be precisely documented, both in docstrings and in the docs. 

Thanks for the nice contribution!

**Comment by farizrahman4u** [comment]
@fchollet Cool! Will be ready tomorrow.

**Comment by farizrahman4u** [comment]
Moved  to #1080

**Comment by Chanrom** [comment]
@farizrahman4u Like in your example of calculating the similarity of two sentences, might there be a problem if my model's first layer is a shard layer? See in the function add_shared_layer at line ```input_layers = [l.layers[-1] for l in inputs]```:
```python
   def add_shared_layer(layer, inputs):
    '''                                                                                                                                                                            
    Use this function to add a shared layer across multiple Sequential models without merging the outputs                                                                          
    '''
      input_layers = [l.layers[-1] for l in inputs]
      s = Siamese(layer, input_layers, merge_mode=None)
      for i in range(len(inputs)):
          sh = SiameseHead(i)
          inputs[i].add(s)
          inputs[i].add(sh)
```

**Comment by farizrahman4u** [comment]
Yes, there is an issue when the first layer of  a model is a shared layer. An  easy fix would be:

```python
   def add_shared_layer(layer, inputs):
    '''                                                                                                                                                                            
    Use this function to add a shared layer across multiple Sequential models without merging the outputs                                                                          
'''
      for i inputs:
          if len(i.layers) == 0:
              i.add(MaskedLayer())
      input_layers = [l.layers[-1] for l in inputs]
      s = Siamese(layer, input_layers, merge_mode=None)
      for i in range(len(inputs)):
          sh = SiameseHead(i)
          inputs[i].add(s)
          inputs[i].add(sh)
```

**Comment by Chanrom** [comment]
Yeah, thank you! 
Thanks for the nice contribution! It helps me a lot.

**Comment by farizrahman4u** [comment]
Glad to help!

**Comment by lxmeng** [comment]
if the first layer is to be shared, adding a maskedlayer can not solve the problem, report:
Exception: Layer is not connected. Did you forget to set "input_shape"?

**Comment by farizrahman4u** [comment]
@lxmeng see #1421

**Comment by lxmeng** [comment]
@farizrahman4u according to your implementation, i can compile and run my siamese-like framework successfully, thx much, but compared with my caffe-implementation, i found that the result has a big gap, have u ever considered to give a siamese example on mnist data like caffe doing?

**Comment by farizrahman4u** [comment]
@lxmeng There is no additional math  involved in Siamese layer.. I simple passed the incoming tensors through the same layer. So if there is a gap, the solution is most probably in the optimizer or the model itself..

**Comment by McVilla** [comment]
Hi, @farizrahman4u Can I use the Siamese layer right now? I can't find it in the master branch. Is it deleted from the master branch or is this PR never merged into master branch? Thanks.

**Comment by fchollet** [comment]
You might want to read the guide "getting started with the functional API"
on keras.io.

**Comment by McVilla** [comment]
@fchollet , I've already read it. I know there is a guide about shared layers. Do you mean I can just use the code example in the guide to construct a Siamese network rather than the example that farizrahman4u showed above?

**Comment by farizrahman4u** [comment]
@McVilla Siamese layer is deprecated. Use functional API instead.
